1-ECommerce/src/components/Menu.tsx

```diff
@@ -0,0 +1,98 @@
+import { useState } from "react";
+
+import { MenuIcon } from "~/ui/Icons";
+import { MenuOption } from "./MenuOption";
+import { Sidebar } from "./Sidebar";
+
+
+export const Menu = () => {
+    const [isMenuVisible, setIsMenuVisible] = useState(false)
+    const [selectedOption, setSelectedOption] = useState(0)
+
+    const handleMenuClick = ()=> {
+        setIsMenuVisible(true)
+    }
+
+    const handleCloseMenu = () => {
+        setSelectedOption(0)
+        setIsMenuVisible(false)
+    }
+
+    const handleMenuOptionClick = (optionId: number) => {
+        const newOptionValue = (selectedOption === optionId) ? 0 : optionId;
+        setSelectedOption(newOptionValue)
+    }
+
+    const options = [
+        { id: 1, name: 'New In'},
+        { id: 2, name: 'Clothing'},
+        { id: 3, name: 'Footwear'},
+        { id: 4, name: 'Accesories'},
+        { id: 5, name: 'SALE'}
+    ]
+
+    const subMenuOptions = [
+        { id: 11, name: 'New In' },
+        { id: 12, name: 'See All'},
+        { id: 13, name: 'Coats'},
+        { id: 14, name: 'Beach clothes'},
+        { id: 15, name: 'Sweaters & hoodies'},
+        { id: 16, name: 'Shirts'},
+        { id: 17, name: 'Jeans & pants'},
+        { id: 18, name: 'T-shirts'},
+        { id: 19, name: 'Shorts'},
+        { id: 20, name: 'Underwear'},
+        { id: 21, name: 'SALE'}
+    ]
+
+    const menuOptions = options
+        .map(option =>
+                <MenuOption
+                    key={option.id}
+                    active={selectedOption === option.id}
+                    option={option}
+                    onOptionClick={() => handleMenuOptionClick(option.id)}
+                />
+    )
+
+    const hardcodedSubMenuOptions = subMenuOptions
+        .map(option =>
+                <MenuOption
+                    key={option.id}
+                    active={false}
+                    option={option}
+                    onOptionClick={() => alert(`suboption "${option.name}" clicked`)}
+                />
+    )
```

Comment:
All of this can be moved outside of the component so it isn't re-declared on every re-render.

Reply:
Great! I think `menuOptions` cannot be moved outside because `handleMEnuOptionClick` is defined inside the component and edits the state of a variable. But the rest definitely belong outside

Reply:
Ah true, you're right